documentation/design/user-stories/shared-infra.adoc

```diff
@@ -744,13 +778,96 @@ spec:
+* If the observed value of these parameters go below min, the number of router replicas is decreased.
+
+```
+apiVersion: enmasse.io/v1beta2
+kind: MessagingInfra
+metadata:
+  name: myinfra
+spec:
+  router:
+    scalingStrategy:
+      dynamic:
+        # Specify the range of replicas this strategy can operate within
+        replicas:
+          min: 2
+          max: 5
+        # Strategy parameters that is used for determining if it should scale up/down
+        parameters:
+          # Number of connections _per router replica_
+          connections:
+            min: 100
+            max: 1000
+          # Number of messages/sec _per router replica_
+          messageThroughputSeconds
+            min: 1000.0
+            max: 15000
+```
 
 === Configure scaling of the brokers based on the queue and topic memory and storage usage
 
+Create/modify the `MessagingInfra` resource to include the `dynamic` strategy for the broker. This strategy scales
+the number of broker instances based on the parameters `addresses` per broker.
+
+* If the observed value of these parameters go above the max, the number of broker instances is increased.
+* If the observed value of these parameters go below min, the number of broker instances is decreased.
+
+```
+apiVersion: enmasse.io/v1beta2
+kind: MessagingInfra
+metadata:
+  name: myinfra
+spec:
+  broker:
+    scalingStrategy:
+      dynamic:
+        # Specify the pool range this strategy can operate within
+        poolSize:
+          min: 2
+          max: 5
+        # Strategy parameters that is used for determining if it should scale up/down
+        parameters:
+          # Number of addresses per broker
+          addresses:
+            min: 100
+            max: 1000
+```
+
 === Configure per-user limits for connections, links, message size
 
-=== Migrate tenant configuration and data from 1.5 to 2.0 (with downtime)
+TODO: Needs support in router plugin and auth service
+
+=== Migrate tenant configuration and data from 0.x to 1.0 (with downtime)
+
+Configuration migration:
+
+* StandardInfraConfig/BrokeredInfraConfig -> MessagingInfra: Manual
+* A separate cluster must be used for the 1.0 cluster
+* AddressPlan -> MessagingAddressPlan: Manual
+* AddressSpacePlan -> MessagingPlan: Manual
+* AddressSpace -> MessagingTenant, MessagingConnector and MessagingEndpoint: Script-assisted
+* Address -> MessagingAddress: Script-assisted
+
+Data migration:
+
+Once the second cluster is configured with tenants and addresses, the next steps are:
+
+* Create user on 0.x cluster with access to all address spaces and addresses
+* Create user on 1.0 cluster with access to all tenants and their addresses
+* Stop traffic to 0.x cluster
+* Run migration application with configured users. The application will:
+** Lookup output data from the address space and address configuration conversion scripts
+** Consume messages from each address in each address space on the 0.x cluster
```

Comment:
Only works for queue types.  Contents of subscriptions would  be lost, unless we made special upgrade rules that allow subscriptions to be routed to as if they were queues.

Reply:
Good point. Maybe a subscription type addresses could have a property specifying if an incoming linkroute/autoLink in the route should be created for it for this purpose? Would it be safer/better to make this a global setting for all subscriptions, settable on the 'MessagingInfra'?

Reply:
If we were to do this, I don't think it should be a standard user-level feature.  It should be just some switch we can throw for the upgrade and then turn it off after the upgrade is complete.  Maybe an annotation?   OTOH maybe we make subscriptions out of scope?

Reply:
If we can avoid caring about subscriptions, that would be my preference.

I agree it would be an annotation and not something we put into the schema. Maybe just an env variable to the operator.